pkg/controller/servingcert/controller/secret_updating_controller.go

```diff
@@ -69,70 +72,141 @@ func (sc *serviceServingCertUpdateController) Key(namespace, name string) (metav
-		return false, nil
-	}
-
-	if sharedService.Annotations[api.ServingCertSecretAnnotation] != secret.Name && sharedService.Annotations[api.AlphaServingCertSecretAnnotation] != secret.Name {
-		return false, nil
-	}
-	if secret.Annotations[api.ServiceUIDAnnotation] != string(sharedService.UID) && secret.Annotations[api.AlphaServiceUIDAnnotation] != string(sharedService.UID) {
-		return false, nil
+		utilruntime.HandleError(fmt.Errorf("unable to get service %s/%s: %v", sharedSecret.Namespace, serviceName, err))
+		return nil
 	}
+	return service
+}
 
+func (sc *serviceServingCertUpdateController) requiresRegenerationFirstCheck(service *v1.Service, secret *v1.Secret, minTimeLeft time.Duration) bool {
 	// if we don't have an ownerref, just go ahead and regenerate.  It's easier than writing a
 	// secondary logic flow.
-	if !ocontroller.HasOwnerRef(secret, ownerRef(sharedService)) {
-		return true, sharedService
+	if !ocontroller.HasOwnerRef(secret, ownerRef(service)) {
+		return true
 	}
-
 	// if we don't have the annotation for expiry, just go ahead and regenerate.  It's easier than writing a
 	// secondary logic flow that creates the expiry dates
 	expiryString, ok := secret.Annotations[api.ServingCertExpiryAnnotation]
 	if !ok {
 		expiryString, ok = secret.Annotations[api.AlphaServingCertExpiryAnnotation]
 		if !ok {
-			return true, sharedService
+			return true
 		}
 	}
 	expiry, err := time.Parse(time.RFC3339, expiryString)
 	if err != nil {
-		return true, sharedService
+		return true
 	}
 
 	if time.Now().Add(sc.minTimeLeftForCert).After(expiry) {
-		return true, sharedService
+		return true
 	}
 
-	return false, nil
+	return false
+}
+
+func (sc *serviceServingCertUpdateController) regenerateSecondCheck(service *v1.Service, secretCopy *v1.Secret) (bool, error) {
+	update := false
+	// If not regenerating, perform checks here to
+	// 1. ensure only the 2 data keys we want exist, TLSCertKey and TLSPrivateKeyKey
+	// 2. ensure cert data is at least a parseable certificate, if not replace with valid data
+	// This does not ensure that somebody didn't swap out secret data for another valid cert.
+	_, ok := secretCopy.Data[v1.TLSCertKey]
+	_, ok2 := secretCopy.Data[v1.TLSPrivateKeyKey]
+	if ok && ok2 {
+		// Don't update, replace unless len data is not 2
+		// This takes care of case where someone added an extra data field. Remove all but 2 keys
+		if len(secretCopy.Data) != 2 {
+			// don't regenerate, just get rid of bad data
+			secretCopy.Data = map[string][]byte{
+				v1.TLSCertKey:       secretCopy.Data[v1.TLSCertKey],
+				v1.TLSPrivateKeyKey: secretCopy.Data[v1.TLSPrivateKeyKey],
+			}
+			update = true
+		}
+	} else {
+		// if TLSCertKey or TLSPrivateKeyKey missing, replace with valid data
+		// Regenerate the secret
+		if err := toRequiredSecret(sc.dnsSuffix, sc.ca, service, secretCopy); err != nil {
+			return false, err
+		}
+		update = true
```

Comment:
right, done.

Reply:
it's good to check, though?, in case someone replaced the cert data with nonsense
(this won't prevent the case where someone for some reason replaces cert data with another valid-ish x509.Cert (ie, it's at least a parseable cert) but will prevent someone from replacing cert data with acsii art)

Reply:
yes, right  I should have, I separated toRequiredSecret from ensureSecretData now, look ok?